rfcs/0008-transition-frontier-controller.md

```diff
@@ -0,0 +1,337 @@
+
+Particularly, we validate transitions with the following checks (in this order):
+
+1. Checking for existence in the breadcrumb tree (we've already seen it)
+2. Checking for consensus errors
+3. Verifying the included SNARK
+
+This order is chosen because it's cheaper computationally to perform (1) and
+(2). So if we get lucky we can short-circuit before getting to the SNARK step.
+
+#### Processor
+
+* Input: `External_transition.t` (from validator) and `Breadcrumb.t` (from catchup)
+* Outputs: `External_transition.t` (to ledger catchup); modifying transition frontier
+
+The processor receives a single validated external transitions from the
+validator and then attempts to add the transition to the breadcrumb tree. The
+processor is the only "thread" allowed to write changes to the
+[Transition Frontier](#transition-frontier). All other threads must delegate to
+the processor.
+
+[Ledger Catchup](#ledger-catchup) needs to add a batch of transitions all at
+once, so it shares the add section and constructs the underlying breadcrumb wired up that can just be $O(1)$ added to the frontier tree.
+
+#### Adding to Transition Frontier
+
+The add process runs in two phases:
+
+1. Perform a `lookup` on the `Transition_frontier` for the previous `State_hash.t` of this transition. If it is absent, send to [catchup monitor](#catchup-monitor). If present, continue.
+2. Apply the `Ledger_builder_diff.t` to the `Breadcrumb.t` retrieved from (1). Note that this no longer mutates the underlying ledger-builder but rather constructs a new mask on top of the existing mask -- see [Transition Frontier](#transition-frontier) for more.
+3. Construct the new `Breadcrumb.t` from the new mask and transition, and attempt a true mutate-add to the underlying [Transition Frontier](#transition-frontier) data.
+
+<a href="catchup-monitor"></a>
+#### Catchup Monitor
+
+The catchup monitor is responsible for waiting a bit before initiating a long
+catchup job. The idea here is to mitigate out-of-order messages since it is much
+quicker to avoid such catchups if possible. This will be a module that waits on
+a small timeout before yielding to the ledger catchup component. And can be
+preempted by some transition/breadcrumb that allows this transition to be
+connected to the existing tree.
+
+<a href="transition-frontier"></a>
+### Transition Frontier
+
+Input: Synchronous writes of `Breadcrumb.t` (coming soley from [Processor](#processor)); Lookup queries coming from all components
+Outputs: None
+
+The Transition Frontier is essentially a `Breadcrumb.t` with metadata at the root of the tree. The root has a merkle database corresponding to the snarked ledger. This is needed for consensus with proof-of-stake. It then has the first mask layer exposing the staged-database and ledger-builder for the locked-tip. Each subsequent breadcrumb contains a light-weight ledger-builder with a masked merkle database built off of the breadcrumb prior.
```

Comment:
It has a root ledger, and a table of breadcrumbs. It just keeps references to the root and best tip in the form of hashes.